lib/sudo-exec/src/signal.rs

```diff
@@ -0,0 +1,46 @@
+use std::{collections::HashMap, ffi::c_int, io};
+
+use signal_hook::consts::*;
+use sudo_system::{
+    poll::PollSet,
+    signal::{SignalHandler, SignalInfo},
+};
+
+pub(crate) const SIGNALS: &[c_int] = &[
+    SIGINT, SIGQUIT, SIGTSTP, SIGTERM, SIGHUP, SIGALRM, SIGPIPE, SIGUSR1, SIGUSR2, SIGCHLD,
+    SIGCONT, SIGWINCH,
+];
+
+pub(crate) struct SignalHandlers {
+    handlers: HashMap<c_int, SignalHandler>,
+    poll_set: PollSet<c_int>,
+}
+
+impl SignalHandlers {
+    pub(crate) fn new() -> io::Result<Self> {
+        let mut handlers = HashMap::with_capacity(SIGNALS.len());
+        let mut poll_set = PollSet::new();
+        for &signal in SIGNALS {
+            let handler = SignalHandler::new(signal)?;
+            poll_set.add_fd_read(signal, &handler);
+            handlers.insert(signal, handler);
+        }
+
+        Ok(Self { handlers, poll_set })
+    }
+
+    pub(crate) fn get_mut(&mut self, signal: c_int) -> Option<&mut SignalHandler> {
+        self.handlers.get_mut(&signal)
+    }
+
+    pub(crate) fn poll(&mut self) -> io::Result<Vec<SignalInfo>> {
+        let signals = self.poll_set.poll()?;
+        let mut infos = Vec::with_capacity(signals.len());
+
+        for signal in signals {
+            infos.push(self.get_mut(signal).unwrap().recv()?);
```

Comment:
something that caught my eye here is that for each signal type `recv` is called *once* but I think that in theory it's possible for a signal (of the same type) to arrive more than once in between `SignalHandlers::poll` calls. in that case, the poll set would report that the signal type is pending but this line would only fetch *one* instance of the signal arrival.

I was thinking that calling `poll` in a loop would prevent that but I'm not sure. does poll detect when a new message comes through the socket or when there's a message that can be read in the socket? if it's the earlier then this issue should be considered, specially for that other `SignalHandlers::pending` method which would only be called once.

Reply:
:thinking: :thought_balloon: 

Yeah we will fetch at most one signal of each kind on a single `SignalHandlers::poll` call even if more signals have arrived. The `poll` syscall will block until of the sockets can be read without blocking. However, we aren't informed how much we can read. 

We can't use [`Read::read_to_end`](https://doc.rust-lang.org/std/io/trait.Read.html#method.read_to_end) instead of `recv` and count how many `SignalInfo` values fit in the size of the buffer because we only get EOF if the other side of the socket disconnects.

Reply:
I added a paragraph to the `recv` documentation specifying this behavior as I think it could be important in an upcoming PR. This is ok for the current event loop as we will either process those signals sooner or later or the command will exit in which case receiving those signals is no longer necessary